### PR TITLE
Tests: testing rejections/errors workflow

### LINKED ISSUE #57: Tests: testing rejections/errors workflow
The majority of tests are about success scenarios and very few are about the rejections and errors that are triggered.

### LINKED ISSUE #66: Stax: ensure ERR_TOO_DEEP/ERR_TOO_LARGE handled correctly

### PR DESCRIPTION
**Pull request by spalmer25**
### Mandatory tests
fixes #57
fixes #66 
blocked by #58 
 - [x] rejection
 - [x] class exception
 - [x] instruction exception
 - [x] p1 (index) exception
 - [x] p2 (derivation type) exception
 - [x] data length exception
 - [x] wrong data
   - [x] wrong derivation path
   - [x] wrong operation (parse error)
   
### Supplementary tests
- Mixing commands: to detect whether a command is being sent while another is in progress.

### CODE REVIEW

File: tests/integration/nanos/test_reject_transaction.sh [hunk 1]
```diff
@@ -0,0 +1,30 @@
+# full input: 0300000000000000000000000000000000000000000000000000000000000000006c016e8874874d31c3fbd636e924d5a036a43ec8faa7d0860308362d80d30e01000000000000000000000000000000000000000000ff02000000020316
+# full output: CAR
+# signer: tz1dyX3B1CFYa2DfdFLyPtiJCfQRUgPVME6E
+# path: m/44'/1729'/0'/0'
+start_speculos "$seed"
+expect_full_text 'ready for' 'safe signing'
+send_async_apdus \
+	800f000011048000002c800006c18000000080000000 "expect_apdu_return 9000
+"\
```
Comment (emturner) [inline tests/integration/nanos/test_reject_transaction.sh:9]: could I ask why this slightly odd formatting (multiple places)?

would `"expect_apdu_return 9000" \` work?

File: app/src/app_main.c [hunk 2]
```diff
@@ -127,6 +127,7 @@ app_main(void)
     global.step = ST_ERROR;
 
     for (;;) {
+        TZ_PREAMBLE(("void"));
```
Comment (emturner) [inline app/src/app_main.c:130]: feel free to update copyright in this file :)

File: app/src/keys.c [hunk 3]
```diff
@@ -59,7 +59,10 @@ read_bip32_path(bip32_path_t *out, const uint8_t *in, size_t in_size)
     TZ_ASSERT(EXC_WRONG_LENGTH_FOR_INS,
               buffer_read_u8(&cdata, &out->length)
                   && buffer_read_bip32_path(
-                      &cdata, (uint32_t *)&out->components, out->length));
+                      &cdata, (uint32_t *)&out->components, out->length)
+                  // Assert all as been read correctly
```
Comment (emturner) [inline app/src/keys.c:63]:
```
// Assert entire bip32_path consumed
```

File: tests/integration/stax/utils.py [hunk 4]
```diff
@@ -134,6 +136,13 @@ def review_skip_to_signing(self, with_loading=False):
         if with_loading: self.assert_screen("loading_operation")
         self.review.client.resume_ticker()
 
+    @contextmanager
```
Comment (emturner) [inline tests/integration/stax/utils.py:139]: this is really nice! I guess we could re-use this in the other `review_*` functions as well (future PR)

File: tests/integration/stax/snapshots/invalid_tag.png
Comment (emturner) [inline tests/integration/stax/snapshots/invalid_tag.png:1]: this PR will also fix #66 I guess

**Review by emturner — COMMENTED** [review]
Overall LGTM, great work! 

A couple of ideas:

it would be great if we could define the error codes in `test_runtime.sh` maybe? rather than having special hard-coded hex everywhere, would be awesome to have `expect_apdu_return $ERR_*` potentially? might help to make the tests more readable/understandable.

As a follow up, I think we've definitely got to the point where it would be good to abstract these tests for the nanos devices slightly - e.g. if we introduce `expect_home` and change `expect_section_content` to use the `$TARGET` env var internally, we could de-duplicate *a lot* of these tests, to avoid needing to always add three copies

**Comment by spalmer25** [comment]
> it would be great if we could define the error codes in `test_runtime.sh` maybe? rather than having special hard-coded hex everywhere, would be awesome to have `expect_apdu_return $ERR_*` potentially? might help to make the tests more readable/understandable.

Yes, that's a great idea!

> As a follow up, I think we've definitely got to the point where it would be good to abstract these tests for the nano devices slightly - e.g. if we introduce `expect_home` and change `expect_section_content` to use the `$TARGET` env var internally, we could de-duplicate _a lot_ of these tests, to avoid needing to always add three copies

It's an excellent idea, but I'll deal with it at another time: #67

File: tests/integration/nanos/test_reject_transaction.sh (hunk 1 above)
Comment (spalmer25) [reply tests/integration/nanos/test_reject_transaction.sh:9]: Yes, the 2 work perfectly.
Some of the tests were written using the test generator, which should have this offset.
I'll deal with this in another MR: #68

**Review by emturner — APPROVED** [review]

**Pull request merged**